ouroboros-consensus/src/Ouroboros/Storage/VolatileDB/Impl.hs

```diff
@@ -458,120 +463,157 @@ reOpenFile :: forall m h blockId
+        withFile hasFS file (AppendMode AllowExisting) $ \hndl ->
+            hTruncate hasFS hndl (fromIntegral offset)
+
+    -- | For each file in the db, this function parses, updates the
+    -- internal state and calls itself for the rest of the files.
+    go :: Index blockId
+       -> ReverseIndex blockId
+       -> SuccessorsIndex blockId
+       -> Maybe (blockId, SlotNo)
+       -> [(FileId, FsPath, FileSize)] -- ^ Info of files with < n blocks.
+       -> [(FileId, FsPath)]
+       -> m (InternalState blockId h)
+    go currentMap currentRevMap succMp _maxSlot lessThanN [] = do
+        hndl <- hOpen hasFS fileToWrite (AppendMode AllowExisting)
+        return $ InternalState {
+            _currentWriteHandle = hndl
+          , _currentWritePath   = fileToWrite
+          , _currentWriteId     = fdToWrite
+          , _currentWriteOffset = offset'
+          , _nextNewFileId      = nextNewFileId'
+          , _currentMap         = currentMap'
+          , _currentRevMap      = currentRevMap
+          , _currentSuccMap     = succMp
+          , _currentMaxSlotNo   = FileInfo.maxSlotInFiles
+                                    (Index.elems currentMap')
+          }
+      where
+        (fileToWrite, fdToWrite, nextNewFileId', currentMap', offset') =
+          case (safeMaximumOn (\(fileId,_,_) -> fileId) lessThanN, lastFile) of
+            (Nothing, _) -> newFileInfo currentMap $ case lastFile of
+              Just (lastFd, _) -> lastFd + 1
+              Nothing          -> 0
+            (_, Nothing) -> error errorMsg
+            (Just (writeFd, wrFile, size), Just (lastFd, _)) | writeFd == lastFd->
+              -- If the file with biggest index is non-empty,
+              -- then this is the file we're looking for.
+              (wrFile, writeFd, lastFd + 1, currentMap, size)
+            (_, Just (lastFd,_)) ->
+              -- If it's not the last file, we just ignore it and
+              -- open a new one.
+              newFileInfo currentMap $ lastFd + 1
+    go currentMap currentRevMap succMap maxSlot lessThanN ((fd, file):rest) = do
+            (blocks, mErr) <- parse parser file
+            updateAndGo blocks mErr
+      where
+        -- | Updates the state and call 'go' for the rest of the files.
+        updateAndGo :: [(SlotOffset, (BlockSize, BlockInfo blockId))]
+                    -> Maybe e
+                    -> m (InternalState blockId h)
+        updateAndGo blocks mErr = do
+            truncateOnError mErr file offset
+            newRevMp <- fromEither err $ reverseMap file currentRevMap fileMp
+            go newMp newRevMp newSuccMp newMaxSlot newHaveLessThanN rest
+          where
+            offset = case reverse blocks of
+              [] -> 0
+              (slotOffset, (blockSize,_)) : _ ->
+                -- The file offset is given by the offset of the last
+                -- block plus its size.
+                slotOffset + blockSize
+            fileMp = Map.fromList blocks
```

Comment:
Use `Map` (like the suggested in my review) instead of `Mp` throughout this function.